Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +21,50 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+   return isCharging_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    if(isCharging_){
+        return timeUntilCharged;
+    }
+    else{
+        return timeUntilDepleted;
+    }
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    if(batteryData.current<0){
+        isCharging_=true;
+    }
+    else{
+        isCharging_=false;
+    }
+
+    if(ampHourUsed==0.0){
+        ampHourUsed=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/PERCENT_TO_DECIMAL);
+    }
+    else{
+        ampHourUsed+=((prevCurr+batteryData.current)/2)*((prevTime.msecsTo(batteryData.time))/MS_HOUR_CONVERSION) ;
```

Comment:
This single line does a lot, I would recommend splitting complex calculation it into multiple ones for readability.